File: src/web/magmac/app/lang/java/JavaAssignmentNode.ts
import { Option } from "../../../../magmac/api/Option";
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { MapNode } from "../../../../magmac/app/compile/node/MapNode";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { LocatingRule } from "../../../../magmac/app/compile/rule/LocatingRule";
import { NodeRule } from "../../../../magmac/app/compile/rule/NodeRule";
import { OrRule } from "../../../../magmac/app/compile/rule/OrRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { Destructors } from "../../../../magmac/app/lang/Destructors";
import { Assignables } from "../../../../magmac/app/lang/node/Assignables";
import { StructureStatementValue } from "../../../../magmac/app/lang/node/StructureStatementValue";
import { TypescriptLang } from "../../../../magmac/app/lang/web/TypescriptLang";
export class JavaAssignmentNode {
	deserialize(node : Node) : Option<CompileResult<JavaAssignmentNode>> {return Destructors.destructWithType( "assignment", node).map( 0);;}
	createAssignmentRule(definition : Rule, value : Rule) : Rule {break;break;return new TypeRule( "assignment", LocatingRule.First( before, "=", source));;}
	serialize() : Node {return new MapNode( "assignment");;}
}

File: src/java/magmac/app/lang/java/JavaAssignmentNode.java
package magmac.app.lang.java;

import magmac.api.Option;
import magmac.api.collect.list.Lists;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.MapNode;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.LocatingRule;
import magmac.app.compile.rule.NodeRule;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.lang.Destructors;
import magmac.app.lang.node.Assignables;
import magmac.app.lang.node.StructureStatementValue;
import magmac.app.lang.web.TypescriptLang;

public record JavaAssignmentNode(JavaLang.JavaAssignable assignable, JavaLang.JavaValue value) implements JavaFunctionSegmentValue, StructureStatementValue, TypescriptLang.FunctionSegment.Value {
    public static Option<CompileResult<JavaAssignmentNode>> deserialize(Node node) {
        return Destructors.destructWithType("assignment", node).map(deserializer -> deserializer
                .withNode("destination", Assignables::deserializeError)
                .withNode("source", JavaDeserializers::deserializeJavaOrError)
                .complete(tuple -> new JavaAssignmentNode(tuple.left(), tuple.right())));
    }

    public static Rule createAssignmentRule(Rule definition, Rule value) {
        Rule before = new NodeRule("destination", new OrRule(Lists.of(
                definition,
                value
        )));

        Rule source = new NodeRule("source", value);
        return new TypeRule("assignment", LocatingRule.First(before, "=", source));
    }

    @Override
    public Node serialize() {
        return new MapNode("assignment");
    }
}
